File: src/LineOfSightLoss.cpp
#include "ITS.Propagation.ITM/ITM.h"

namespace ITS {
namespace Propagation {
namespace ITM {
    
/*=============================================================================
 |
 |  Description:  Compute the loss in the line-of-sight region
 |
 |        Input:  d__meter          - Path distance, in meters
 |                h_e__meter[2]     - Terminal effective heights, in meters
 |                Z_g               - Complex surface transfer impedance
 |                delta_h__meter    - Terrain irregularity parameter
 |                M_d               - Diffraction slope
 |                A_d0              - Diffraction intercept
 |                d_sML__meter      - Maximum line-of-sight distance for
 |                                    a smooth earth, in meters
 |                f__mhz            - Frequency, in MHz
 |
 |      Outputs:  [None]
 |
 |      Returns:  A_los__db         - Loss, in dB
 |
 *===========================================================================*/
double LineOfSightLoss(double d__meter, double h_e__meter[2], complex<double> Z_g, double delta_h__meter, 
    double M_d, double A_d0, double d_sML__meter, double f__mhz)
{
    double delta_h_d__meter = TerrainRoughness(d__meter, delta_h__meter);

    double sigma_h_d__meter = SigmaHFunction(delta_h_d__meter);

    // wavenumber, k
    double wn = f__mhz / 47.7;

    // [Algorithm, Eqn 4.46]
    double sin_psi = (h_e__meter[0] + h_e__meter[1]) / sqrt(pow(d__meter, 2) + pow(h_e__meter[0] + h_e__meter[1], 2));

    // [Algorithm, Eqn 4.47]
    complex<double> R_e = (sin_psi - Z_g) / (sin_psi + Z_g) * exp(-MIN(10.0, wn * sigma_h_d__meter * sin_psi));

    // q = Magnitude of R_e', [Algorithm, Eqn 4.48]
    double q = pow(R_e.real(), 2) + pow(R_e.imag(), 2);
    if (q < 0.25 || q < sin_psi)
        R_e = R_e * sqrt(sin_psi / q);

    // phase difference between rays, [Algorithm, Eqn 4.49]
    double delta_phi = wn * 2.0 * h_e__meter[0] * h_e__meter[1] / d__meter;

    // [Algorithm, Eqn 4.50]
    if (delta_phi > PI / 2.0)
        delta_phi = PI - pow(PI / 2.0, 2) / delta_phi;

    // Two-ray attenuation
    complex<double> rr = complex<double>(cos(delta_phi), -sin(delta_phi)) + R_e;
    double A_t__db = -10 * log10(pow(rr.real(), 2) + pow(rr.imag(), 2));

    // Extended diffraction attenuation
    double A_d__db = M_d * d__meter + A_d0;

    // weighting factor
    double w = 1 / (1 + f__mhz * delta_h__meter / MAX(10e3, d_sML__meter));

    double A_los__db = w * A_t__db + (1 - w) * A_d__db;

    return A_los__db;
}

}  // namespace ITM
}  // namespace Propagation
}  // namespace ITS

File: itm-rs/itm/src/line_of_sight_loss.rs
use crate::complex::Complex;
use crate::constants::PI;
use crate::sigma_h_function::sigma_h_function;
use crate::terrain_roughness::terrain_roughness;

/// Line-of-sight path loss, in dB.
pub fn line_of_sight_loss(
    d__meter: f64,
    h_e__meter: [f64; 2],
    z_g: Complex,
    delta_h__meter: f64,
    m_d: f64,
    a_d0: f64,
    d_sml__meter: f64,
    f__mhz: f64,
) -> f64 {
    let delta_h_d = terrain_roughness(d__meter, delta_h__meter);
    let sigma_h_d = sigma_h_function(delta_h_d);
    let wn = f__mhz / 47.7;

    let sin_psi =
        (h_e__meter[0] + h_e__meter[1]) / (d__meter.powi(2) + (h_e__meter[0] + h_e__meter[1]).powi(2)).sqrt();

    let exp_val = (-f64::min(10.0, wn * sigma_h_d * sin_psi)).exp();
    let r_e_raw = (sin_psi - z_g) / (sin_psi + z_g) * exp_val;

    let q = r_e_raw.re.powi(2) + r_e_raw.im.powi(2);
    let r_e = if q < 0.25 || q < sin_psi {
        r_e_raw * (sin_psi / q).sqrt()
    } else {
        r_e_raw
    };

    let mut delta_phi = wn * 2.0 * h_e__meter[0] * h_e__meter[1] / d__meter;
    if delta_phi > PI / 2.0 {
        delta_phi = PI - (PI / 2.0).powi(2) / delta_phi;
    }

    let rr = Complex::new(delta_phi.cos(), -delta_phi.sin()) + r_e;
    let a_t = -10.0 * (rr.re.powi(2) + rr.im.powi(2)).log10();

    let a_d = m_d * d__meter + a_d0;
    let w = 1.0 / (1.0 + f__mhz * delta_h__meter / f64::max(10_000.0, d_sml__meter));

    w * a_t + (1.0 - w) * a_d
}
